File: src/main/java/org/flighthub/Service/Service.java
package org.flighthub.Service;
import org.flighthub.Domain.*;
import org.flighthub.Repository.*;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

public class Service {
    private AgentRepository agentRepository;
    private ClientRepository clientRepository;
    private FlightRepository flightRepository;
    private TicketRepository ticketRepository;
    private TouristRepository touristRepository;

    public Service(AgentRepository agentRepository, ClientRepository clientRepository, FlightRepository flightRepository, TicketRepository ticketRepository,TouristRepository touristRepository) {
        this.agentRepository = agentRepository;
        this.clientRepository = clientRepository;
        this.flightRepository = flightRepository;
        this.ticketRepository = ticketRepository;
        this.touristRepository=touristRepository;
    }

    //Login the Agent
    public Agent login(String username, String password){
        Agent agent = agentRepository.login(username,password);
        return agent;
    }
    //Save one Agent
    public void saveAgent(String username,String passwword){
        Agent agent=new Agent(username);
        agentRepository.save(agent,passwword);
    }
    //Get all Flights - not showing flights with no available seats
    public List<Flight> getFlights(){
        return filterFlights((List<Flight>) flightRepository.findAll());
    }
    //Search flight by destination & date
    public List<Flight> searchFlight(String destination, LocalDate departureDate){
        return filterFlights((List<Flight>) flightRepository.findByDestinationAndDepartureDate(destination,departureDate));
    }
    //Buy ticket
    public boolean buyTicket(String clientName,String clientAddress,String touristsName,int seats,Flight flight){
        try{
        Client client=new Client(clientName,clientAddress);
        clientRepository.save(client);
        String[] touristsNameArray = touristsName.split(", ");
        List<Tourist> touristList=new ArrayList<>();
        for(String name:touristsNameArray){
            Tourist tourist=new Tourist(name);
            touristRepository.save(tourist);
            touristList.add(tourist);
        }
        Ticket ticket=new Ticket(client,touristList,flight,seats);
        ticketRepository.save(ticket);
        return true;}
        catch (Exception exception){
            return false;
        }
    }
    // Filter flights
    public List<Flight> filterFlights(List<Flight> flights) {
        List<Flight> filteredFlights = new ArrayList<>();
        for (Flight flight : flights) {
            if (flight.getAvailableSeats() > 0) {
                filteredFlights.add(flight);
            }
        }
        return filteredFlights;
    }
    //Add entities
    public void addEntities() {
        Agent agent = new Agent("TudorCalin");
        agentRepository.save(agent, "password");

        Client client1 = new Client("Alice Johnson", "123 Main St, Springfield, IL 62701, USA");
        clientRepository.save(client1);

        Client client2 = new Client("Bob Smith", "456 Elm St, New York, NY 10001, USA");
        clientRepository.save(client2);

        Flight flight1 = new Flight("New York", LocalDateTime.of(2024, 3, 15, 10, 0), 200, "JFK Airport");
        flightRepository.save(flight1);

        Flight flight2 = new Flight("Los Angeles", LocalDateTime.of(2024, 4, 20, 12, 30), 150, "LAX Airport");
        flightRepository.save(flight2);

        Flight flight3 = new Flight("Chicago", LocalDateTime.of(2024, 5, 10, 9, 15), 180, "ORD Airport");
        flightRepository.save(flight3);

        Flight flight4 = new Flight("Miami", LocalDateTime.of(2024, 6, 5, 8, 45), 220, "MIA Airport");
        flightRepository.save(flight4);

        Flight flight5 = new Flight("Seattle", LocalDateTime.of(2024, 7, 18, 14, 20), 190, "SEA Airport");
        flightRepository.save(flight5);

        Flight flight6 = new Flight("San Francisco", LocalDateTime.of(2024, 8, 22, 11, 10), 170, "SFO Airport");
        flightRepository.save(flight6);

        Flight flight7 = new Flight("Las Vegas", LocalDateTime.of(2024, 9, 30, 16, 0), 160, "LAS Airport");
        flightRepository.save(flight7);

        Flight flight8 = new Flight("Orlando", LocalDateTime.of(2024, 10, 12, 13, 45), 210, "MCO Airport");
        flightRepository.save(flight8);

        Tourist tourist1 = new Tourist("Emily");
        touristRepository.save(tourist1);

        Tourist tourist2 = new Tourist("Michael");
        touristRepository.save(tourist2);
    }

}

File: Server - C#/Server/Service.cs
using chat.services;
using FlightHubC_.Domain;
using FlightHubC_.Repository;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace FlightHubC_.ServiceNamespace
{
    public class Service : IServices
    {
        private readonly AgentRepository agentRepository;
        private readonly ClientRepository clientRepository;
        private readonly FlightRepository flightRepository;
        private readonly TicketRepository ticketRepository;
        private readonly TouristRepository touristRepository;

        private readonly IDictionary<String, IObserver> loggedClients;


        public Service(AgentRepository agentRepository, ClientRepository clientRepository, FlightRepository flightRepository, TicketRepository ticketRepository, TouristRepository touristRepository)
        {
            this.agentRepository = agentRepository;
            this.clientRepository = clientRepository;
            this.flightRepository = flightRepository;
            this.ticketRepository = ticketRepository;
            this.touristRepository = touristRepository;
            loggedClients = new Dictionary<String, IObserver>();

        }

        // Login the Agent
        public Agent Login(string username, string password)
        {
            return agentRepository.Login(username, password);
        }

        // Save one Agent
        public void SaveAgent(string username, string password)
        {
            Agent agent = new Agent(username);
            agentRepository.Save(agent, password);
        }

        // Get all Flights - not showing flights with no available seats
        public List<Flight> getFlights()
        {
            return (List<Flight>)flightRepository.FindAll();
        }

        // Search flight by destination & date
        public List<Flight> SearchFlight(string destination, DateTime departureDate)
        {
            return (List<Flight>)flightRepository.FindByDestinationAndDepartureDate(destination, departureDate);
        }

        // Buy ticket
        public bool BuyTicket(string clientName, string clientAddress, string touristsName, int seats, Flight flight)
        {

            Client client = new Client(clientName, clientAddress);
            clientRepository.Save(client);

            string[] touristsNameArray = touristsName.Split(',');
            List<Tourist> touristList = new List<Tourist>();
            foreach (string name in touristsNameArray)
            {
                Tourist tourist = new Tourist(name);
                touristRepository.save(tourist);
                touristList.Add(tourist);
            }

            Ticket ticket = new Ticket(client, touristList, flight, seats);
            ticketRepository.Save(ticket);
            NotifyClientsTicketBought();
            return true;

        }
        public Flight findByDetails(string destination, DateTime departureDateTime, int availableSeats, string airport)
        {
            return flightRepository.FindByDetails(destination, departureDateTime, availableSeats, airport);
        }

        // Add entities
        public void AddEntities()
        {
            Agent agent = new Agent("TudorCalin");
            agentRepository.Save(agent, "password");

            Client client1 = new Client("Alice Johnson", "123 Main St, Springfield, IL 62701, USA");
            clientRepository.Save(client1);

            Flight flight1 = new Flight("New York", DateTime.Now.AddDays(1), 100, "NY Airport");

            Flight flight2 = new Flight("Los Angeles", DateTime.Now.AddDays(2), 150, "LA Airport");
            flightRepository.Save(flight2);

            Flight flight3 = new Flight("London", DateTime.Now.AddDays(3), 200, "Heathrow Airport");
            flightRepository.Save(flight3);

            Flight flight4 = new Flight("Paris", DateTime.Now.AddDays(4), 120, "Charles de Gaulle Airport");
            flightRepository.Save(flight4);

            Flight flight5 = new Flight("Tokyo", DateTime.Now.AddDays(5), 180, "Narita International Airport");
            flightRepository.Save(flight5);

            Flight flight6 = new Flight("Dubai", DateTime.Now.AddDays(6), 250, "Dubai International Airport");
            flightRepository.Save(flight6);

            Flight flight7 = new Flight("Sydney", DateTime.Now.AddDays(7), 180, "Sydney Airport");
            flightRepository.Save(flight7);
            
        }
        public Flight findFlightId(String id)
        {
            return flightRepository.FindOne(Guid.Parse(id));
        }

        public void logout(Agent agent, IObserver client)
        {
            loggedClients.Remove(agent.Id.ToString());
        }


        public Agent login(string username, string password, IObserver client)
        {
            Agent agent = agentRepository.Login(username,password);
            if (agent != null)
            {
               
                loggedClients[agent.Id.ToString()] = client;
                return agent;
            }
            else
                throw new TException("Authentication failed.");
        }
        public void NotifyClientsTicketBought()
        {
            foreach(var observer in loggedClients.Values)
            {
                Task.Run(() => observer.notifyAgents());
            }
        }
    }
}
